policy/handler.go

```diff
@@ -141,6 +156,61 @@ func (h *Handler) Stop() {
 	h.running = false
 }
 
+func (h *Handler) handleTick(ctx context.Context, policy *Policy) (*Evaluation, error) {
+
+	// Timestamp the invocation of this evaluation run. This can be
+	// used when checking cooldown or emitting metrics to ensure some
+	// consistency.
+	curTime := time.Now().UTC().UnixNano()
+
+	eval, err := h.generateEvaluation(policy)
+	if err != nil {
+		return nil, err
+	}
+
+	// If the evaluation is nil there is nothing to be done this time
+	// around.
+	if eval == nil {
+		return nil, nil
+	}
+
+	// If the target status includes a last event meta key, check for cooldown
+	// due to out-of-band events. This is also useful if the Autoscaler has
+	// been re-deployed.
+	ts, ok := eval.TargetStatus.Meta[targetpkg.MetaKeyLastEvent]
+	if !ok {
+		return eval, nil
+	}
+
+	// Convert the last event string. If an error occurs, just log and
+	// continue with the evaluation. A malformed timestamp shouldn't mean
+	// we skip scaling.
+	lastTS, err := strconv.ParseUint(ts, 10, 64)
+	if err != nil {
+		h.log.Error("failed to parse last event timestamp as uint64", "error", err)
+		return eval, nil
+	}
+
+	// Check whether we need to enter cooldown or not.
+	if !h.isInCooldown(policy.Cooldown, curTime, lastTS) {
+		return eval, nil
+	}
+
+	// Calculate the remaining time period left on the cooldown.
+	cdPeriod := h.calculateRemainingCooldown(policy.Cooldown, curTime, int64(lastTS))
+
+	// Enforce the cooldown which will block until complete. A false response
+	// means we did not reach the end of cooldown due to a request to shutdown.
+	if !h.enforceCooldown(ctx, cdPeriod) {
+		return nil, fmt.Errorf("cooldown period interupted by shutdown request")
```

Comment:
This will cause an error message to be log, which might be confusing and distract operators during a debug session, but we need to return an error here so the main loop terminates without risking reading from any of the other channels.

We could use a custom error type here (or maybe `context.Canceled`) and check for it in the main loop to avoid logging.